Added functionality to Discord input

ajCameron opened the pull request:
Added
- Input events when messages are edited and deleted (not just created)
- Input events when a new member joins a guild
- The ability to request historic messages on startup
- Example bots for all these (apart from the new member join functionality - need a good way to test this)
- Refactored the Discord Input event class names for consistency
- Refactored examples to isolate all the Discord ones in a single subfolder

ajCameron committed:
Adding on_member_join event to io.discord

 - Added DiscordUserJoinInputEvent
- Which is emitted whenever a new user joins any of the guilds the bot is monitoring

ajCameron committed:
Added trivial load test

- Corrected problem with blacking
- Added a load test for DiscordIO

ajCameron committed:
Added DiscordMessageEditInputEvent

Triggers when there's an edit to a message in any of the channels the bot is monitoring.
 - Produces an input event with the original message and the new one

ajCameron committed:
Added example for "full" discord bot

- Edit aware discord bot example added
- Refactoring to add more Discord Input event types
- Old DiscordInputEvent not longer sufficiently precise - renamed to DiscordTextInputEvent
- Example tested and working

ajCameron committed:
Added a queue depth to discord input

- In line with the RSS input, history messages can now be retrieved
- Added example for this functionality
- Gave up on a single example to demonstrate each new feature of the discord input - now going to have a new example for each of them

ajCameron committed:
Refactored example to include a discord_bots subfolder

- Moved all examples
- Refactored in the yaml
- Tested they still work

ajCameron committed:
Implemented DiscordMessageDeleteEvents

- Which are produced whenever a message in a discord guild monitored by the bot is deleted
- Refactored DiscordTextInput to the (more in line with the rest of the classes) DiscordMessageCreationEvent
- Added new example - delete_warn_discord_bot (tested, working)

exachixkitsune commented on examples/discord_bots/delete_warn_discord_bot.py:
```diff
@@ -0,0 +1,106 @@
+
+class DiscordAllCommandTrigger(Trigger):
+    """
+    Nothing fancy - just fires whenever there is a DiscordTextInputEvent - of any type.
```
Probably wants modifying if that's no longer the case? Or does the event trigger, and then it gets used on some instances?

ajCameron replied:
No - I think that's just a stale comment. Will update.

ajCameron replied:
Updated examples and comments to reflect new variable names - some more refactoring of the examples for clarity.

exachixkitsune reviewed (dismissed):
Looks good, only one comment on the comments in the files.
Are there other tests we could add to tests_io_discord.py? It feels like it may be hard/impossible to do a full test in a discord, but can we emulate that behaviour? - not important for this maybe

ajCameron commented:
I've been thinking about how to test Discord properly as well.

I need to finish hacking it together, but I think there's scope for a full test - no mocking required. My approach would be to create a throwaway (hidden) channel, rig the Input class so it ignores all events but the ones in that channel, and then run the tests - emptying the channel after each. (The user would need to feed the tests suite a valid discord token ... but that's probably not too much to as).

Will think on it some more, and might give it a crack.

ajCameron committed:
Refactoring to tidy up examples

- Examples now more clearly demonstrate the capability of the commands
- Refactoring of doc strings to reflect new variable names
- Changed names of example Triggers and Actions to reflect what they actually do

ajCameron commented:
Refactored some of the examples for added clarity (i.e. making the doc strings actually reflect reality).

javajawa commented on src/mewbot/io/discord.py:
```diff
@@ -91,6 +160,54 @@ async def on_ready(self) -> None:
+            past_messages, key=lambda x: float(x.created_at.timestamp()), reverse=True
+        )
+
+        for _ in range(self._startup_queue_depth):
+            message = past_messages[_]
```
I do not like using `_` as an actual variable.
Would something like a slice make this clearer?
```suggestion
        for message in past_messages[0:self._startup_queue_depth]:
```

ajCameron replied:
Clearer and possibly faster!

ajCameron committed:
Refactored to use slice to avoid use of _ as a variable

Co-authored-by: Benedict Harcourt <ben.harcourt@harcourtprogramming.co.uk>
